Xamarin.Forms.Maps.Android/MapRenderer.cs

```diff
@@ -10,311 +11,358 @@
-			for (var i = 0; i < Map.Pins.Count; i++) {
-				var pin = Map.Pins[i];
+			for (var i = 0; i < Map.Pins.Count; i++)
+			{
+				Pin pin = Map.Pins[i];
 				if ((string)pin.Id != marker.Id)
+				{
 					continue;
+				}
 
 				targetPin = pin;
 				break;
 			}
 
 			// only consider event handled if a handler is present. 
 			// Else allow default behavior of displaying an info window.
-			targetPin?.SendTap ();
+			targetPin?.SendTap();
 		}
 
-		bool _disposed;
-		protected override void Dispose (bool disposing)
+		void MoveToRegion(MapSpan span, bool animate)
 		{
-			if (disposing && !_disposed) {
-				_disposed = true;
-			
-				var mapModel = Element as Map;
-				if (mapModel != null) {
-					MessagingCenter.Unsubscribe<Map, MapSpan> (this, MoveMessageName);
-					((ObservableCollection<Pin>)mapModel.Pins).CollectionChanged -= OnCollectionChanged;
+			GoogleMap map = NativeMap;
+			if (map == null)
+			{
+				return;
+			}
+
+			span = span.ClampLatitude(85, -85);
+			var ne = new LatLng(span.Center.Latitude + span.LatitudeDegrees / 2,
+				span.Center.Longitude + span.LongitudeDegrees / 2);
+			var sw = new LatLng(span.Center.Latitude - span.LatitudeDegrees / 2,
+				span.Center.Longitude - span.LongitudeDegrees / 2);
+			CameraUpdate update = CameraUpdateFactory.NewLatLngBounds(new LatLngBounds(sw, ne), 0);
+
+			try
+			{
+				if (animate)
+				{
+					map.AnimateCamera(update);
+				}
+				else
+				{
+					map.MoveCamera(update);
 				}
+			}
+			catch (IllegalStateException exc)
+			{
+				System.Diagnostics.Debug.WriteLine("MoveToRegion exception: " + exc);
+			}
+		}
 
-				var gmap = NativeMap;
-				if (gmap == null)
-					return;
-				gmap.MyLocationEnabled = false;
-				gmap.InfoWindowClick -= MapOnMarkerClick;
-				gmap.Dispose ();
+		void OnCollectionChanged(object sender, NotifyCollectionChangedEventArgs notifyCollectionChangedEventArgs)
+		{
+			switch (notifyCollectionChangedEventArgs.Action)
+			{
+				case NotifyCollectionChangedAction.Add:
+					AddPins(notifyCollectionChangedEventArgs.NewItems);
+					break;
+				case NotifyCollectionChangedAction.Remove:
+					RemovePins(notifyCollectionChangedEventArgs.OldItems);
+					break;
+				case NotifyCollectionChangedAction.Replace:
+					RemovePins(notifyCollectionChangedEventArgs.OldItems);
+					AddPins(notifyCollectionChangedEventArgs.NewItems);
+					break;
+				case NotifyCollectionChangedAction.Reset:
+					_markers?.ForEach(m => m.Remove());
+					_markers = null;
+					AddPins((IList)Element.Pins);
+					break;
+				case NotifyCollectionChangedAction.Move:
+					//do nothing
+					break;
+			}
+		}
+
+		void OnMoveToRegionMessage(Map s, MapSpan a)
+		{
+			MoveToRegion(a, true);
+		}
+
+		void RemovePins(IList pins)
+		{
+			GoogleMap map = NativeMap;
+			if (map == null)
+			{
+				return;
+			}
+			if (_markers == null)
+			{
+				return;
+			}
+
+			foreach (Pin p in pins)
+			{
+				Marker marker = _markers.FirstOrDefault(m => (object)m.Id == p.Id);
```

Comment:
No reason it couldn't be, I suppose. The fact that it's coming from an enumerable called `_markers` makes the type obvious enough to pass [rule 10](https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/coding-style.md).